packages/runtime/src/enhancements/node/encrypted.ts

```diff
@@ -0,0 +1,155 @@
+
+    private async encrypt(field: FieldInfo, data: string): Promise<string> {
+        if (this.isCustomEncryption(this.options.encryption!)) {
+            return await this.options.encryption.encrypt(this.model, field, data);
+        }
+
+        const key = await getKey(this.options.encryption!.encryptionKey);
+        const iv = crypto.getRandomValues(new Uint8Array(12));
+
+        const encrypted = await crypto.subtle.encrypt(
+            {
+                name: 'AES-GCM',
+                iv,
+            },
+            key,
+            encoder.encode(data)
+        );
+
+        // Combine IV and encrypted data into a single array of bytes
+        const bytes = [...iv, ...new Uint8Array(encrypted)];
+
+        // Convert bytes to base64 string
+        return btoa(String.fromCharCode(...bytes));
+    }
+
+    private async decrypt(field: FieldInfo, data: string): Promise<string> {
+        if (this.isCustomEncryption(this.options.encryption!)) {
+            return await this.options.encryption.decrypt(this.model, field, data);
+        }
+
+        const key = await getKey(this.options.encryption!.encryptionKey);
+
+        // Convert base64 back to bytes
+        const bytes = Uint8Array.from(atob(data), (c) => c.charCodeAt(0));
+
+        // First 12 bytes are IV, rest is encrypted data
+        const decrypted = await crypto.subtle.decrypt(
+            {
+                name: 'AES-GCM',
+                iv: bytes.slice(0, 12),
+            },
+            key,
+            bytes.slice(12)
+        );
+
+        return decoder.decode(decrypted);
+    }
+
+    // base override
+    protected async preprocessArgs(action: PrismaProxyActions, args: any) {
+        const actionsOfInterest: PrismaProxyActions[] = ['create', 'createMany', 'update', 'updateMany', 'upsert'];
+        if (args && args.data && actionsOfInterest.includes(action)) {
+            await this.preprocessWritePayload(this.model, action as PrismaWriteActionType, args);
+        }
+        return args;
+    }
+
+    // base override
+    protected async processResultEntity<T>(method: PrismaProxyActions, data: T): Promise<T> {
+        if (!data || typeof data !== 'object') {
+            return data;
+        }
+
+        for (const value of enumerate(data)) {
+            await this.doPostProcess(value, this.model);
+        }
+
+        return data;
+    }
+
+    private async doPostProcess(entityData: any, model: string) {
+        const realModel = this.queryUtils.getDelegateConcreteModel(model, entityData);
+
+        for (const field of getModelFields(entityData)) {
+            const fieldInfo = await resolveField(this.options.modelMeta, realModel, field);
+
+            if (!fieldInfo) {
+                continue;
+            }
+
+            const shouldDecrypt = fieldInfo.attributes?.find((attr) => attr.name === '@encrypted');
+            if (shouldDecrypt) {
+                entityData[field] = await this.decrypt(fieldInfo, entityData[field]);
```

Comment:
If decryption fails, should we return the original cipher text? I'm thinking this will allow easier adoption: the `@encrypted` attribute can be added and deployed and then a background script is run to migrate the existing plain-text data.

Reply:
That may be vulnerable to data leaks some edge cases I think. For example, if custom encryption is used and the encrypt function fails resulting in storing the data in plain text, then it would leak during read operations.

I'm thinking that an encrypted field should be treated as all or nothing.

what do you think?

Reply:
That may be vulnerable to data leaks for some edge cases I think. For example, if custom encryption is used and the encrypt function fails resulting in storing the data in plain text, then it would leak during read operations.

I'm thinking that an encrypted field should be treated as all or nothing. (e.g. throw an error if encrypt fails, or return a null of decrypt fails?)

what do you think?

Reply:
I agree encryption should be strict and the write operation should just fail as a whole if encrypting a field fails.

However, I'm worried treating decryption "strict" will make incremental adoption very difficult. If one already has some rows and then turn on `@encrypted`, for a period of time, the column will be mixed with plain and cipher text, unless the system is brought down to maintenance to allow a migration process to complete.

As long as the encryption process's integrity is guaranteed, if decryption fails (e.g., due to incorrect key etc.), the worst case scenario will be cipher text is returned the the caller, right?